src/gtk/dc.cpp

```diff
@@ -237,6 +240,76 @@ void* wxGTKCairoDCImpl::GetCairoContext() const
     return cr;
 }
 
+bool wxGTKCairoDCImpl::TryGetWindowSize(wxSize& size, wxSize& sizeMM) const
+{
+    if ( !m_window )
+        return false;
+
+    GdkWindow* const w = gtk_widget_get_window(m_window->GetHandle());
+    if ( !w )
+        return false;
+
+#if GTK_CHECK_VERSION(3, 22, 0)
+    if ( wx_is_at_least_gtk3(22) )
+    {
+        GdkDisplay* const d = gdk_window_get_display(w);
+        if ( !d )
+            return false;
+
+        GdkMonitor* const m = gdk_display_get_monitor_at_window(d, w);
+        if ( !m )
+            return false;
+
+        GdkRectangle rect;
+        gdk_monitor_get_geometry(m, &rect);
+
+        // The returned geometry is in the application pixels, convert them to
+        // the real device pixels to obtain the real PPI.
+        const int scale = gdk_monitor_get_scale_factor(m);
+        size.x = rect.width * scale;
+        size.y = rect.height * scale;
+
+        sizeMM.x = gdk_monitor_get_width_mm(m);
+        sizeMM.y = gdk_monitor_get_height_mm(m);
+
+        return true;
+    }
+#endif // GTK+ >= 3.22
+
+    // For GTK+4 the code above is always used, but for GTK+ 3 we fall back to
+    // it when run-time GTK+ version is < 3.22 or when compiled with < 3.22 in
+    // the first place.
+#ifndef __WXGTK4__
+    wxGCC_WARNING_SUPPRESS(deprecated-declarations)
+    GdkScreen* const s = gdk_window_get_screen(w);
+    if ( !s )
+        return false;
+
+    size.x = gdk_screen_get_width(s);
+    size.y = gdk_screen_get_height(s);
+    wxGCC_WARNING_RESTORE()
+#endif // GTK+ < 4
+
+    return true;
+}
+
+wxSize wxGTKCairoDCImpl::GetPPI() const
+{
+    wxSize size, sizeMM;
+    if ( TryGetWindowSize(size, sizeMM) )
+    {
+        return wxSize(wxRound((inches2mm * size.x) / sizeMM.x),
+                      wxRound((inches2mm * size.y) / sizeMM.y));
```

Comment:
I did it like this to have a multiplication in the numerator rather than denominator, but, thinking more about this, it probably doesn't really matter and the compiler will likely do the right thing anyhow, so I'll make it more clear, thanks.